### PR TITLE
Currency reject support

### PR DESCRIPTION
**Pull request by dhutsalo**
added logic for currency rejection similar to Go version. Default currency is USD

### CODE REVIEW

File: src/main/java/org/prebid/server/auction/ExchangeService.java [hunk 1]
```diff
@@ -672,20 +673,47 @@ private static ExtPriceGranularity parsePriceGranularity(JsonNode priceGranulari
      * <p>
      * Returns input argument as the result if no errors found or create new {@link BidderSeatBid} otherwise.
      */
-    private BidderSeatBid validateAndUpdateResponse(BidderSeatBid bidderSeatBid) {
+    private BidderSeatBid validateAndUpdateResponse(BidderSeatBid bidderSeatBid, List<String> requestCurrencies) {
+        if (requestCurrencies.size() == 0) {
```
Comment (schernysh) [inline src/main/java/org/prebid/server/auction/ExchangeService.java]: Please extract seatbid currency validation and bids validation into separate methods which `validateAndUpdateResponse` will call.

File: src/test/java/org/prebid/server/auction/ExchangeServiceTest.java [hunk 2]
```diff
@@ -1602,6 +1602,32 @@ public void shouldUpdatePriceForOneBidAndDropAnotherIfPrebidExceptionHappensForS
                 entry("bidder", singletonList("no currency conversion available")));
     }
 
+    @Test
+    public void shouldResponseWithErrorWhenRequestsWithUnsupportedCurrency()
```
Comment (schernysh) [inline src/test/java/org/prebid/server/auction/ExchangeServiceTest.java]: Typo, should read `shouldRespond...WhenBidsWith...`

File: src/test/java/org/prebid/server/auction/ExchangeServiceTest.java [hunk 3]
```diff
@@ -1768,6 +1794,11 @@ private static BidRequest givenBidRequest(
         return bidRequestBuilderCustomizer.apply(BidRequest.builder().cur(singletonList("USD")).imp(imp)).build();
     }
 
+    private static BidRequest givenBidRequestWithCur(
```
Comment (schernysh) [inline src/test/java/org/prebid/server/auction/ExchangeServiceTest.java]: I would argue that creation of this method is justified - it is used only once so `givenBidRequest` could apply here.

File: src/main/java/org/prebid/server/auction/ExchangeService.java [hunk 4]
```diff
@@ -672,20 +673,47 @@ private static ExtPriceGranularity parsePriceGranularity(JsonNode priceGranulari
...
         final List<BidderError> errors = new ArrayList<>(bidderSeatBid.getErrors());
 
-        for (BidderBid bid : bids) {
-            final ValidationResult validationResult = responseBidValidator.validate(bid.getBid());
-            if (validationResult.hasErrors()) {
-                for (String error : validationResult.getErrors()) {
-                    errors.add(BidderError.generic(error));
+        if (bids.size() > 0) {
+            //assume that currencies are the same among all bids
+            String bidsCurrency = bids.get(0).getBidCurrency();
```
[the remark this answers is not in the document]
Comment (dhutsalo) [reply src/main/java/org/prebid/server/auction/ExchangeService.java]: looks like not
 ObjectUtils.firstNonNull(bids.toArray(new BidderBid[]{})) will return `null` for empty collection that will cause NPE

File: src/test/java/org/prebid/server/auction/ExchangeServiceTest.java (hunk 3 above)
Comment (dhutsalo) [reply src/test/java/org/prebid/server/auction/ExchangeServiceTest.java]: considering future improvements to currency support, `givenBidRequest` can stand for BidRequest with default currency

File: src/main/java/org/prebid/server/auction/ExchangeService.java [hunk 5]
```diff
@@ -672,20 +673,47 @@ private static ExtPriceGranularity parsePriceGranularity(JsonNode priceGranulari
...
+                    validCurencyPresent = true;
+                    break;
                 }
+            }
+
+            if (!validCurencyPresent) {
+                String requestCursFormated = requestCurrencies.stream()
+                        .collect(Collectors.joining(",", "", ""));
+                errors.add(BidderError.generic(String.format(
+                        "%s is not allowed. Use %s instead", requestCursFormated, bidsCurrency)));
```
[the remark this answers is not in the document]
Comment (dhutsalo) [reply src/main/java/org/prebid/server/auction/ExchangeService.java]: used error like in go verison

File: src/main/java/org/prebid/server/auction/ExchangeService.java [hunk 6]
```diff
@@ -672,12 +672,88 @@ private static ExtPriceGranularity parsePriceGranularity(JsonNode priceGranulari
...
+        return validBids.size() == bids.size()
+                ? bidderSeatBid
+                : BidderSeatBid.of(validBids, bidderSeatBid.getHttpCalls(), errors);
+    }
+
+    private boolean isValidCurFromResponse(List<String> requestCurrencies, List<BidderBid> bids,
+                                           List<BidderError> errors) {
+        if (bids.size() > 0) {
+            //assume that currencies are the same among all bids
+            String bidsCurrency = nonNullCurrencyOtherwiseDefault(bids.get(0).getBidCurrency());
```
Comment (schernysh) [inline src/main/java/org/prebid/server/auction/ExchangeService.java]: Please don't forget to make local variables final

File: src/main/java/org/prebid/server/auction/ExchangeService.java [hunk 7]
```diff
@@ -672,12 +672,88 @@ private static ExtPriceGranularity parsePriceGranularity(JsonNode priceGranulari
...
+
+            boolean sameCurrenciesAcrossAllBids = isValidCurrencyPresent(bids);
```
Comment (schernysh) [inline src/main/java/org/prebid/server/auction/ExchangeService.java]: Variable could be inlined

File: src/main/java/org/prebid/server/auction/ExchangeService.java [hunk 8]
```diff
@@ -672,12 +672,88 @@ private static ExtPriceGranularity parsePriceGranularity(JsonNode priceGranulari
...
+
+            if (!sameCurrenciesAcrossAllBids) {
+                errors.add(BidderError.generic("Bid currencies mismatch found. "
+                        + "Expected all bids to have the same currencies."));
+                return false;
+            }
+
+            boolean validCurrencyPresent = isValidCurrencyPresent(requestCurrencies, bidsCurrency);
```
Comment (schernysh) [inline src/main/java/org/prebid/server/auction/ExchangeService.java]: Variable could be inlined

File: src/main/java/org/prebid/server/auction/ExchangeService.java [hunk 9]
```diff
@@ -672,12 +672,88 @@ private static ExtPriceGranularity parsePriceGranularity(JsonNode priceGranulari
...
+
+            if (!validCurrencyPresent) {
+                final String requestCursFormatted = requestCurrencies.stream()
```
Comment (schernysh) [inline src/main/java/org/prebid/server/auction/ExchangeService.java]: `String.join` is less verbose for this case, also this variable could be inlined for better readability

File: src/main/java/org/prebid/server/auction/ExchangeService.java [hunk 10]
```diff
@@ -672,12 +672,88 @@ private static ExtPriceGranularity parsePriceGranularity(JsonNode priceGranulari
...
+                currency = localCurrency;
+            } else if (!currency.equals(localCurrency)) {
+                sameCurrenciesAcrossAllBids = false;
+            }
+        }
+
+        return sameCurrenciesAcrossAllBids;
+    }
+
+    private boolean isValidCurrencyPresent(List<String> requestCurrencies, String bidsCurrency) {
```
Comment (schernysh) [inline src/main/java/org/prebid/server/auction/ExchangeService.java]: Please replace this method with `List.contains` method

File: src/main/java/org/prebid/server/auction/ExchangeService.java [hunk 11]
```diff
@@ -672,12 +672,88 @@ private static ExtPriceGranularity parsePriceGranularity(JsonNode priceGranulari
...
+                        requestCursFormatted, bidsCurrency)));
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    private boolean isValidCurrencyPresent(List<BidderBid> bids) {
+        boolean sameCurrenciesAcrossAllBids = true;
```
Comment (schernysh) [inline src/main/java/org/prebid/server/auction/ExchangeService.java]: This method could be simplified using streams: `bids.stream().map(map to currency).distinct().count() ==1`

File: src/test/java/org/prebid/server/auction/ExchangeServiceTest.java (hunk 3 above)
Comment (schernysh) [reply src/test/java/org/prebid/server/auction/ExchangeServiceTest.java]: Let's remove this unused method

File: src/main/java/org/prebid/server/auction/ExchangeService.java (hunk 10 above)
Comment (dhutsalo) [reply src/main/java/org/prebid/server/auction/ExchangeService.java]: replace it with smth like this : 
`bids.stream().map(BidderBid::getBidCurrency).collect(Collectors.toList()).contains(bids.get(0).getBidCurrency())`

File: src/main/java/org/prebid/server/auction/ExchangeService.java [hunk 12]
```diff
@@ -673,12 +673,52 @@ private static ExtPriceGranularity parsePriceGranularity(JsonNode priceGranulari
...
+        return validBids.size() == bids.size()
+                ? bidderSeatBid
+                : BidderSeatBid.of(validBids, bidderSeatBid.getHttpCalls(), errors);
+    }
+
+    private boolean isValidCurFromResponse(List<String> requestCurrencies, List<BidderBid> bids,
+                                           List<BidderError> errors) {
+        if (bids.size() > 0) {
+            //assume that currencies are the same among all bids
+            final String bidsCurrency = nonNullCurrencyOtherwiseDefault(bids.get(0).getBidCurrency());
```
Comment (schernysh) [inline src/main/java/org/prebid/server/auction/ExchangeService.java]: I would say this line and the next `if` could be combined into something like:
```
final List<String> bidderCurrencies = bids.stream()
    .map(bid -> ObjectUtils.firstNonNull(bid.getBidCurrency(), DEFAULT_CURRENCY))
    .distinct()
    .collect(Collectors.toList());

if(bidderCurrencies.size > 1) {
    errors....
    return false;
}

if (!requestCurrencies.contains(bidderCurrencies.get(0))) {
...
```

File: src/main/java/org/prebid/server/auction/ExchangeService.java [hunk 13]
```diff
@@ -695,18 +695,21 @@ private boolean isValidCurFromResponse(List<String> requestCurrencies, List<Bidd
...
 
-            if (bids.stream().map(BidderBid::getBidCurrency).distinct().count() != 1) {
+            if (bidderCurrencies.size() > 1) {
                 errors.add(BidderError.generic("Bid currencies mismatch found. "
                         + "Expected all bids to have the same currencies."));
                 return false;
             }
 
-            if (!requestCurrencies.contains(bidsCurrency)) {
+            if (!requestCurrencies.contains(bidderCurrencies.get(0))) {
```
Comment (schernysh) [inline src/main/java/org/prebid/server/auction/ExchangeService.java]: Let's extract `bidderCurrencies.get(0)` to a variable

**Pull request merged**